crypto/provider_core.c

```diff
@@ -107,25 +116,41 @@ static struct provider_store_st *get_provider_store(OPENSSL_CTX *libctx)
         return NULL;
 
     store = openssl_ctx_get_data(libctx, provider_store_index);
-    if (store == NULL)
+    if (store == NULL) {
         CRYPTOerr(CRYPTO_F_GET_PROVIDER_STORE, ERR_R_INTERNAL_ERROR);
+    } else {
+        CRYPTO_THREAD_write_lock(store->lock);
+        if (sk_OSSL_PROVIDER_num(store->providers) == 0) {
+            const struct predefined_providers_st *p = NULL;
+
+            for (p = predefined_providers; p->name == NULL; p++) {
+                OSSL_PROVIDER *prov = NULL;
+
+                if (p->name == NULL)
+                    continue;
+                /*
+                 * We use the internal constructor directly here,
+                 * otherwise we get a call loop
+                 */
+                prov = provider_new(p->name, p->init);
+
+                if (prov == NULL
+                    || sk_OSSL_PROVIDER_push(store->providers, prov) == 0) {
+                    ossl_provider_free(prov);
+                    CRYPTOerr(CRYPTO_F_GET_PROVIDER_STORE,
+                              ERR_R_INTERNAL_ERROR);
+                    return NULL;
+                }
+                prov->store = store;
+                if(p->is_fallback)
+                    ossl_provider_set_fallback(prov);
+            }
+        }
+        CRYPTO_THREAD_unlock(store->lock);
```

Comment:
So every time we do a fetch or touch the store we will obtain a write lock for a short time? I think this could be a big source of lock contention. But actually this code only ever needs to be done once - at the time the store is created. Wouldn't it be better placed in `provider_store_new` to avoid the lock?

Reply:
True.  Ok, hold on, will experiment with that

Reply:
Done